### PR TITLE
ZA | 25-SDC-July | Luke Manyamazi | Sprint 3 | Implement Shell Tools Exercises

### LINKED ISSUE #176: 01 Implement shell tools (cat, ls, wc) in JavaScript with NodeJS
### Link to the coursework

https://github.com/CodeYourFuture/Module-Tools/tree/main/implement-shell-tools

You should implement these tools in JavaScript.

### Why are we doing this?

Practicing designing and writing code, getting used to writing small tools, and understanding argument parsing.

### Maximum time in hours

9

### How to submit

Submit a PR from your fork of the Module-Tools repo with your implementations.

### PR DESCRIPTION
**Pull request by Luke-Manyamazi**
## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)
- [x] My changes meet the [requirements](./README.md) of this task

## Changelist

This PR includes implementations of basic Unix-like command-line tools in Node.js:

ls: Lists files in a directory.
cat: Reads and prints the contents of one or more files.
wc: Counts lines, words, and characters in one or more files, with support for flags (-l, -w, -c).

Each command was tested to match expected output behavior for the given use cases.

## Questions

1. Is my handling of default behavior in the wc command (i.e. when no flags are provided) aligned with how the actual Unix wc works?
2. Do you recommend any improvements for structuring output formatting, especially when dealing with multiple files and totals?
3. Are there any edge cases you suggest I test further?
4.

### CODE REVIEW

File: implement-shell-tools/cat/cat.mjs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    const fileStat = await stat(file);
+    if (fileStat.isDirectory()) {
+      console.error(`cat: ${file}: Is a directory`);
+      continue;
+    }
+    const content = await readFile(file, 'utf-8');
+    const lines = content.split('\n');
+
+    for (const line of lines) {
+      const shouldNumber = (numberAllLines && !numNotBlank) || (numNotBlank && line.trim() !== '');
```
Comment (LonMcGregor) [inline implement-shell-tools/cat/cat.mjs:28]: You are not numbering non-blank lines when -b is used. Can you see why that is?

File: implement-shell-tools/ls/ls.mjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { program } from "commander"
+import process from "node:process"
+import { promises as fs } from "node:fs"
+import { readdir } from 'node:fs/promises'
+
+program
+    .name("ls")
+    .description("Lists the files in a directory")
+    .option("-1, --one", "One per line")
+    .option("-a", "Include files starting with dot")
```
Comment (LonMcGregor) [inline implement-shell-tools/ls/ls.mjs:10]: You offer -a as an option, but is this ever used?

File: implement-shell-tools/wc/wc.mjs
Comment (LonMcGregor) [inline implement-shell-tools/wc/wc.mjs:1]: Throughout this file there is quite a lot of duplication. Can you think of how you might reduce this?

**Review by LonMcGregor — COMMENTED** [review]
Good work on this, I've left some comments for where you could start to improve more.

For your questions, yes, it looks like you're handling defaults for wc fine. I think you are handling output fine.

If you are thinking of handling more edge cases, think about how people might use these commands - what sort of arguments might get passed to your code?

**Review by LonMcGregor — COMMENTED** [review]
Thanks for making these changes. cat and ls look good now. I still notice some duplication in the wc file however.

Have a look at the last 2 blocks of code. They're both doing the same thing: Check options, push to an array,add padding, then print out. Is there a way you could abstract that code a bit?

**Review by LonMcGregor — COMMENTED** [review]
Looks great! You're finished with this task now

**Comment by illicitonion** [comment]
Closing PR because the SDC run has finished. Feel free to re-open if you're still working on it.
